### ISSUE TITLE
panic: runtime error: invalid memory address or nil pointer dereference

### ISSUE THREAD

**Issue by toast-uz** [eclipse/paho.mqtt.golang#115]
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x38 pc=0x5f78d7]

goroutine 3007 [running]:
github.com/eclipse/paho%2emqtt%2egolang.alllogic(0xc420ba5c00)
	/home/centos/projects/src/src/github.com/eclipse/paho.mqtt.golang/net.go:265 +0x2297
created by github.com/eclipse/paho%2emqtt%2egolang.(*client).Connect.func1
	/home/centos/projects/src/src/github.com/eclipse/paho.mqtt.golang/client.go:265 +0xac0
exit status 2

I tried benchmark for mosquitto using paho.mqtt.golang and this error occurred always around 200k-300k msgs published from paho to mosquitto by QoS=2. No problem occurred on the same situation by QoS=0 or 1.

**Comment by titpetric** [comment]
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x1b265c]

goroutine 20 [running]:
app/vendor/github.com/eclipse/paho%2emqtt%2egolang.alllogic(0x106ba000)
        /go/src/app/vendor/github.com/eclipse/paho.mqtt.golang/net.go:265 +0x18f4
created by app/vendor/github.com/eclipse/paho%2emqtt%2egolang.(*client).Connect.func1
        /go/src/app/vendor/github.com/eclipse/paho.mqtt.golang/client.go:265 +0x7f8
```

Seems the issue I ran into is identical. Will degrade to QOS=1 and see if it improves.

**Comment by titpetric** [comment]
Also, duplicate of #112, PR in https://github.com/eclipse/paho.mqtt.golang/pull/113

**Comment by alsm** [comment]
#113 resolves the effect of a runtime error, but it's hiding the problem that somehow we are getting back a nil for a messageID that we think is still live. I'd really like to resolve that one otherwise you'll potentially end up with a token.Wait() waiting indefinitely because we never think it's completed.
If someone has a simple test case of code and setup that can be used to reproduce this I'd like to try and get to the bottom of it.

**Comment by titpetric** [comment]
Having an underlying issue is not a good reason not to fix the obvious issue. I now vendored the changes from PR 133, which is unfortunate. I'd also give a good stab at rewriting a lot of the client code here, if Eclipse would sponsor some of this work? :)

**Comment by alsm** [comment]
The nil check is not a fix, it's a mask, somehow we've lost the underlying token and without knowing why we could have leaks and unpredictable behaviour if something else is waiting on that token, and spotting that happening will be harder and potentially more disruptive.

**Comment by pshirali** [comment]
@alsm , this is the same issue as #93 (which was reported on an older version).
I now have a chance to spend some time on it. I'll keep you posted.

**Comment by alsm** [comment]
Thanks, my latest commit does something similar to the changes proposed, but I check for the nil token at look up time in messageids.go and if we get a nil token from a lookup return a dummy token which will print out the messageID it was requested for when flowComplete() is called.

**Comment by pshirali** [comment]
@alsm - I was able to repro once so far: `A lookup for token 4 returned nil`

**Comment by alsm** [comment]
As part of reverting the changes to the array rather than the map I've left in the changes that will return a dummy token in the case we can't find one in the lookup table, this will prevent the panics while I look further into why we're losing tokens/looking up incorrect message ids.

**Comment by alsm** [comment]
I'm going to close this issue in favour of opening one specifically for the missing/incorrect tokens/ids.
